File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const N = maps.length, M = maps[0].length;
    const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    let visited = Array.from({length: N}, () => Array(M).fill(false));
    visited[0][0] = true;
    
    let queue = [[0, 0, 1]];
    let head = 0;
    while(head < queue.length){
        const [x, y, moveDistance] = queue[head++];
        
        if(x === N-1 && y === M-1) return moveDistance;
        
        for([dx, dy] of directions){
            const mx = x + dx, my = y + dy;
            if(0 <= mx && mx < N && 0 <= my && my < M && maps[mx][my] === 1 && !visited[mx][my]){
                visited[mx][my] = true;
                queue.push([mx, my, moveDistance + 1]);
            }
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    N, M = len(maps), len(maps[0])
    visited = [[False] * M for _ in range(N)]
    visited[0][0] = True
    
    directions = [[-1, 0], [1, 0], [0, 1], [0, -1]]
    q = deque([[0, 0, 1]])
    while q:
        x, y, distance = q.popleft()
        if x == N-1 and y == M-1:
            return distance
        for dx, dy in directions:
            mx, my = x + dx, y + dy
            if 0 <= mx < N and 0 <= my < M and maps[mx][my] == 1 and not visited[mx][my]:
                visited[mx][my] = True
                q.append([mx, my, distance+1])
    
    return -1